### ISSUE TITLE
callback for the optimized variable in inverse problem

### ISSUE THREAD

**Issue by sashahexe** [lululxvi/deepxde#395]
Dear Lu Lu,

could you please give me a hint. 

1) I am compiling the model first with adam and then with L-BFGS-B. I put the `dde.callbacks.VariableValue()` before the final `losshistory, train_state` parts but the file for the variable gets updated only once after adam is trained and never gets next optimized values from the  L-BFGS-B. Could you please help me what is there to change in the code?

`   model = dde.Model(data, net)

    model.compile('adam', lr=0.001, external_trainable_variables=[gamma])

    model.train(epochs=5000)   

    model.compile("L-BFGS-B", external_trainable_variables=[gamma])   
    
    model.train_step.optimizer_kwargs = {'options': {'maxfun': 1e5, 'ftol': 1e-20, 'gtol': 1e-20, 'eps': 1e-20, 'iprint': -1, 'maxiter': 1e5}} 

    variable = dde.callbacks.VariableValue([gamma], period=1000, filename="gamma_2.dat")

    losshistory, train_state = model.train(epochs=50000, callbacks=[variable]) `

Here is my code https://colab.research.google.com/drive/18I6pRggrKomVuJ2ZIM0OchrIPt_2YePH?usp=sharing

Thank you in advance for your time!

**Comment by Pranay-Nagrani** [comment]
Hi Dr. Lulu,

Does this mean while solving an inverse problem to identify unknown model parameters, I can only use Adam optimizer to train my model and not L-BFGS-B?

Thanks in advance!

**Comment by lululxvi** [comment]
You could. But you can only show the variable values at the end.

**Comment by Pranay-Nagrani** [comment]
Hi Dr. Lulu,

I train my model first with Adam and then using L-BFGS-B. My train and test loss reduces when I optimize with L-BFGS-B but model parameters are the same as the last iteration of Adam. Am I missing something? The snippet of my code where I implement L-BFGS-B is:

model.compile("L-BFGS-B",external_trainable_variables=[G, eta_p, yield_stress])
variable = dde.callbacks.VariableValue([G, eta_p, yield_stress], period=10, filename="variables_1.dat")
losshistory, trainstate = model.train(callbacks = [variable])
dde.saveplot(losshistory, trainstate, issave=True, isplot=True)

**Comment by lululxvi** [comment]
The code seems good.

**Comment by okhsallam** [comment]
How can I know the last value for the variable trained by the LBFGS  ?

**Comment by lululxvi** [comment]
@okhsallam See code above.

**Comment by tsarikahin** [comment]
@lululxvi I think it is not called at the end but at the beginning. 

model.py line 419

```python
        self.callbacks.on_train_begin()
        if optimizers.is_external_optimizer(self.opt_name):
            if backend_name == "tensorflow.compat.v1":
                self._train_tensorflow_compat_v1_scipy(display_every)
            elif backend_name == "tensorflow":
                self._train_tensorflow_tfp()
            elif backend_name == "pytorch":
                self._train_pytorch_lbfgs()
        else:
            if epochs is None:
                raise ValueError("No epochs for {}.".format(self.opt_name))
            self._train_sgd(epochs, display_every)
        self.callbacks.on_train_end()
```
The `on_train_end` function does not exits in VariableValue class. Therefore the unknown variable is output only at the very beginning. 

@Pranay-Nagrani @okhsallam

**Comment by AndersNormannHermanrud** [comment]
I used this as a workaround at the end, after L-BFGS-B training. C.txt should be a new file so the old one is not overwritten.
```
variable = dde.callbacks.VariableValue(C, period=1, filename="C.txt")
model.compile("adam", lr=0, external_trainable_variables=C)
losshistory, train_state = model.train(epochs=0, callbacks=[variable], display_every=1)
```

**Comment by tsarikahin** [comment]
This could be a trick or at the source code one can add the following code block in callback.py after `on_train_begin` function 

```python
def on_train_end(self):
        if backend_name == "tensorflow.compat.v1":
            self.value = self.model.sess.run(self.var_list)
        elif backend_name == "tensorflow":
            self.value = [var.numpy() for var in self.var_list]
        elif backend_name == "pytorch":
            self.value = [var.detach().item() for var in self.var_list]
        print(
            self.model.train_state.epoch,
            list_to_str(self.value, precision=self.precision),
            file=self.file,
        )
        self.file.flush()
```
